### PR TITLE
Switch from using sum for flattening lists of lists in group_texts

### PR DESCRIPTION
**Pull request by nbroad1881**
# Speed up list flattening in `group_texts`  by changing `sum(list_of_lists, [])` to `functools.reduce(operator.iconcat, list_of_lists, [])`

I changed all list flattening from `sum(list_of_lists, [])` to `functools.reduce(operator.iconcat, list_of_lists, [])`. 

Here is a stack overflow thread about which method is fastest: https://stackoverflow.com/a/45323085

Here is a colab notebook that shows a quick example between the old way and the new way and a couple of timed examples. The new way is about 5-6x faster. https://colab.research.google.com/drive/1Kxj_JbM9HMLFpjUduy6i3tfqDob_pYIp?usp=sharing

I discovered this while trying to use `group_texts` on many GB of data, and the speedup was greatly appreciated. 

Nearly all of these changes are in `run_mlm` or `run_clm` examples, but there are a couple in `run_swag` and another 
 in `file_utils.py` which might be unnecessary.

I don't know why `make style` is moving `import functools` to its own line above the other imports in examples/flax/language-modeling/run_t5_mlm_flax.py and examples/tensorflow/language-modeling/run_clm.py

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

I think @sgugger wrote the original `group_texts`

### CODE REVIEW

**Comment by sgugger** [comment]
Thanks for investigating the performance of that line! I initially used the `sum()` because it's easy to read. The `functools.reduce(operator.iconcat, list_of_lists, [])` is way less readable, but it gives so much in speed that switching may be a good thing.

Wdyt @LysandreJik ?

**Comment by nbroad1881** [comment]
I was actually confused at first because I didn't know that `sum()` could be used to flatten a list of lists like that. Perhaps just adding a comment explaining that it is flattening the list of lists into one list will suffice?

Alternatively, would `chain.from_iterable(list_of_lists)` or `chain(*list_of_lists)` seem more readable?

**Comment by sgugger** [comment]
I think `chain(*list_of_lists)` is the most readable, and is even clearer than the `sum` thing.
It makes almost no difference in time from your notebook, so let's go with this one?

**Comment by nbroad1881** [comment]
Ok I'll make the changes. Do you know why `make style` moved `import functools` into its own line in `examples/flax/language-modeling/run_t5_mlm_flax.py` and `examples/tensorflow/language-modeling/run_clm.py`?

**Review by LysandreJik — APPROVED** [review]
This looks good to me ! The `chain(*examples)` is indeed way clearer than `functools.reduce(operator.iconcat, list_of_lists, [])` and `sum(examples, [])`.

Thanks for looking into it, @nbroad1881!

File: examples/flax/language-modeling/run_clm_flax.py [hunk 1]
```diff
@@ -21,12 +21,14 @@
 """
 # You can also adapt this script on your own causal language modeling task. Pointers for this are left as comments.
 
+
```
Comment (sgugger) [inline examples/flax/language-modeling/run_clm_flax.py:24]:
```suggestion
```

File: examples/flax/language-modeling/run_mlm_flax.py [hunk 2]
```diff
@@ -20,11 +20,13 @@
 Here is the full list of checkpoints on the hub that can be fine-tuned by this script:
 https://huggingface.co/models?filter=masked-lm
 """
+
```
Comment (sgugger) [inline examples/flax/language-modeling/run_mlm_flax.py:23]:
```suggestion
```

File: examples/flax/language-modeling/run_t5_mlm_flax.py [hunk 3]
```diff
@@ -19,12 +19,15 @@
 Here is the full list of checkpoints on the hub that can be pretrained by this script:
 https://huggingface.co/models?filter=t5
 """
+
+
```
Comment (sgugger) [inline examples/flax/language-modeling/run_t5_mlm_flax.py:23]:
```suggestion
```

File: examples/pytorch/language-modeling/run_clm.py [hunk 4]
```diff
@@ -21,11 +21,13 @@
 """
 # You can also adapt this script on your own causal language modeling task. Pointers for this are left as comments.
 
+
```
Comment (sgugger) [inline examples/pytorch/language-modeling/run_clm.py:24]:
```suggestion
```

File: examples/pytorch/multiple-choice/run_swag.py [hunk 5]
```diff
@@ -18,10 +18,12 @@
 """
 # You can also adapt this script on your own multiple choice task. Pointers for this are left as comments.
 
+
```
Comment (sgugger) [inline examples/pytorch/multiple-choice/run_swag.py:21]:
```suggestion
```

File: examples/tensorflow/language-modeling/run_clm.py [hunk 6]
```diff
@@ -22,6 +22,7 @@
 """
 # You can also adapt this script on your own clm task. Pointers for this are left as comments.
 
+
```
Comment (sgugger) [inline examples/tensorflow/language-modeling/run_clm.py:25]:
```suggestion
```

**Review by sgugger — APPROVED** [review]
LGTM as well, thanks for amending your PR! Let's just remove all those blank new lines before merging.

**Comment by nbroad1881** [comment]
I did a couple more tests in this notebook: https://colab.research.google.com/drive/1Kxj_JbM9HMLFpjUduy6i3tfqDob_pYIp

One way to improve readability would be to make a utility function like this: `ravel = functools.partial(functools.reduce, operator.iconcat, [])` so then you could just use `ravel(x)` inside `group_texts`.  `ravel` would mirror what `torch.ravel` and `np.ravel` do. I'm not sure if that is more or less confusing.

This method actually performed the fastest out of everything I tried.

Here is a summary of the methods when using `group_texts` on SQuAD contexts where `x` is a list of lists:

1. `sum(x, [])` - 1 loop, best of 3: 2min 46s per loop
2. `list(chain.from_iterable(x))` - 1 loop, best of 3: 27 s per loop
3. `list(chain(*x))` - 1 loop, best of 3: 27.2 s per loop
4. `functools.reduce(operator.iconcat, x, [])` - 1 loop, best of 3: 26.7 s per loop
5. `functools.partial(functools.reduce, operator.iconcat, [])(x)` - 1 loop, best of 3: 25.8 s per loop
6. `np.ravel(x)` - 1 loop, best of 3: 29.1 s per loop
7. `[b  for a in x for b in a]` - 1 loop, best of 3: 28 s per loop

**Comment by sgugger** [comment]
I think that option 3 (`list(chain(*x))`) is the best compromise in terms of readability vs speed (only 3ms longer than the best run, which might also be dataset-dependent).

Thanks a lot for benchmarking all options!

**Pull request merged**